qlib/contrib/model/pytorch_gru.py

```diff
@@ -0,0 +1,362 @@
+        x_valid = x_valid.fillna(0)
+        y_valid = y_valid.fillna(0)
+        x_train.to_pickle('~/x_train.pkl')
+        y_train.to_pickle('~/y_train.pkl')
+
+        # Lightgbm need 1D array as its label
+        save_path = create_save_path(save_path)
+        stop_steps = 0
+        train_loss = 0
+        best_loss = np.inf
+        evals_result["train"] = []
+        evals_result["valid"] = []
+
+        # train
+        self.logger.info("training...")
+        self._fitted = True
+        # return
+        # prepare training data
+        x_train_values = torch.from_numpy(x_train.values).float()
+        y_train_values = torch.from_numpy(np.squeeze(y_train.values)).float()
+        train_num = y_train_values.shape[0]
+
+        # prepare validation data
+        x_val_auto = torch.from_numpy(x_valid.values).float()
+        y_val_auto = torch.from_numpy(np.squeeze(y_valid.values)).float()
+
+        if self.use_gpu:
+            x_val_auto = x_val_auto.cuda()
+            y_val_auto = y_val_auto.cuda()
+
+        for step in range(self.n_epochs):
+            if stop_steps >= self.early_stop:
+                if verbose:
+                    self.logger.info("\tearly stop")
+                break
+            loss = AverageMeter()
+            self.gru_model.train()
+            self.train_optimizer.zero_grad()
+
+            choice = np.random.choice(train_num, self.batch_size)
+            x_batch_auto = x_train_values[choice]
+            y_batch_auto = y_train_values[choice]
+
+            if self.use_gpu:
+                x_batch_auto = x_batch_auto.float().cuda()
+                y_batch_auto = y_batch_auto.float().cuda()
+
+            # forward
+            preds = self.gru_model(x_batch_auto)
+            cur_loss = self.get_loss(preds, y_batch_auto, self.loss_type)
+            cur_loss.backward()
+            self.train_optimizer.step()
+            loss.update(cur_loss.item())
+
+            # validation
+            train_loss += loss.val
+            # print(loss.val)
+            if step and step % self.eval_steps == 0:
+                stop_steps += 1
+                train_loss /= self.eval_steps
+
+                with torch.no_grad():
+                    self.gru_model.eval()
+                    loss_val = AverageMeter()
+
+                    # forward
+                    preds = self.gru_model(x_val_auto)
+                    cur_loss_val = self.get_loss(preds, y_val_auto, self.loss_type)
+                    loss_val.update(cur_loss_val.item())
+
+                if verbose:
+                    self.logger.info(
+                        "[Epoch {}]: train_loss {:.6f}, valid_loss {:.6f}".format(step, train_loss, loss_val.val)
+                    )
+                evals_result["train"].append(train_loss)
+                evals_result["valid"].append(loss_val.val)
+                if loss_val.val < best_loss:
+                    if verbose:
+                        self.logger.info(
```

Comment:
@Derek-Wds is design an experiment results recorder. 
You can use his tools after he finishes